File: dotnet/FnDsa/src/Hqc/HqcKem.cs
using System.Security.Cryptography;

namespace FnDsa.Hqc;

/// <summary>
/// HQC (Hamming Quasi-Cyclic) Key Encapsulation Mechanism.
/// Provides KeyGen, Encaps, and Decaps operations for CCA-secure KEM.
/// Uses SHAKE-256 (via Shake256Impl) for hashing and seed expansion.
/// </summary>
public static class HqcKem
{
    /// <summary>
    /// Generates an HQC key pair.
    /// </summary>
    /// <returns>(publicKey, secretKey)</returns>
    public static (byte[] pk, byte[] sk) KeyGen(HqcParams p)
    {
        // Generate random seeds
        byte[] skSeed = RandomNumberGenerator.GetBytes(HqcParams.SeedBytes);
        byte[] pkSeed = RandomNumberGenerator.GetBytes(HqcParams.SeedBytes);

        // Generate secret vectors x, y from sk_seed
        var skExpander = new SeedExpander(skSeed);
        ulong[] x = VectSetRandomFixedWeight(skExpander, p.N, p.W);
        ulong[] y = VectSetRandomFixedWeight(skExpander, p.N, p.W);

        // Generate random vector h from pk_seed
        var pkExpander = new SeedExpander(pkSeed);
        ulong[] h = VectSetRandom(pkExpander, p.N);

        // Compute s = x + h * y mod (x^n - 1)
        ulong[] hy = GF2.VectMul(h, y, p.N);
        ulong[] s = GF2.VectAdd(hy, x);
        s = GF2.VectResize(s, p.N);

        // Serialize public key: [pk_seed (40 bytes)] [s (VecNSizeBytes bytes)]
        byte[] pk = new byte[p.PKSize];
        Array.Copy(pkSeed, pk, HqcParams.SeedBytes);
        byte[] sBytes = GF2.VectToBytes(s, p.VecNSizeBytes);
        Array.Copy(sBytes, 0, pk, HqcParams.SeedBytes, sBytes.Length);

        // Serialize secret key: [sk_seed (40 bytes)] [pk (PKSize bytes)]
        byte[] sk = new byte[p.SKSize];
        Array.Copy(skSeed, sk, HqcParams.SeedBytes);
        Array.Copy(pk, 0, sk, HqcParams.SeedBytes, pk.Length);

        return (pk, sk);
    }

    /// <summary>
    /// Encapsulates a shared secret using the public key.
    /// </summary>
    /// <returns>(ciphertext, sharedSecret)</returns>
    public static (byte[] ct, byte[] ss) Encaps(byte[] pk, HqcParams p)
    {
        // Generate random message m
        byte[] m = RandomNumberGenerator.GetBytes(p.VecKSizeBytes);

        // Compute d = H(m)
        byte[] d = ComputeD(m);

        // Compute theta
        byte[] theta = ComputeTheta(m, pk, d, p);

        // PKE Encrypt
        var (u, v) = PkeEncrypt(m, theta, pk, p);

        // Compute shared secret
        byte[] ss = ComputeSS(m, u, v, p);

        // Serialize ciphertext: [u (VecNSizeBytes)] [v (VecN1N2SizeBytes)] [d (64 bytes)]
        byte[] ct = new byte[p.CTSize];
        byte[] uBytes = GF2.VectToBytes(u, p.VecNSizeBytes);
        byte[] vBytes = GF2.VectToBytes(v, p.VecN1N2SizeBytes);
        Array.Copy(uBytes, ct, uBytes.Length);
        Array.Copy(vBytes, 0, ct, p.VecNSizeBytes, vBytes.Length);
        Array.Copy(d, 0, ct, p.VecNSizeBytes + p.VecN1N2SizeBytes, d.Length);

        return (ct, ss);
    }

    /// <summary>
    /// Decapsulates a shared secret from a ciphertext using the secret key.
    /// </summary>
    public static byte[] Decaps(byte[] sk, byte[] ct, HqcParams p)
    {
        // Parse secret key
        byte[] skSeed = sk[..HqcParams.SeedBytes];
        byte[] pk = sk[HqcParams.SeedBytes..];

        // Parse ciphertext
        ulong[] u = GF2.VectFromBytes(ct.AsSpan(0, p.VecNSizeBytes), p.VecNSize64);
        ulong[] v = GF2.VectFromBytes(ct.AsSpan(p.VecNSizeBytes, p.VecN1N2SizeBytes), p.VecN1N2Size64);
        byte[] d = ct[(p.VecNSizeBytes + p.VecN1N2SizeBytes)..];

        // Regenerate secret vectors x, y and sigma from sk_seed
        var skExpander = new SeedExpander(skSeed);
        _ = VectSetRandomFixedWeight(skExpander, p.N, p.W); // x (not needed for decrypt)
        ulong[] y = VectSetRandomFixedWeight(skExpander, p.N, p.W);
        // Generate sigma (rejection secret)
        byte[] sigma = new byte[p.VecKSizeBytes];
        skExpander.Read(sigma);

        // Compute v - u * y (XOR in GF(2))
        ulong[] uy = GF2.VectMul(u, y, p.N);

        // Truncate uy to n1*n2 bits
        var uyTrunc = new ulong[p.VecN1N2Size64];
        Array.Copy(uy, uyTrunc, Math.Min(uy.Length, p.VecN1N2Size64));
        if (p.N1N2 % 64 != 0 && p.VecN1N2Size64 > 0)
            uyTrunc[p.VecN1N2Size64 - 1] &= (1UL << (p.N1N2 % 64)) - 1;

        ulong[] vMinusUY = GF2.VectAdd(v, uyTrunc);

        // Decode using tensor product code
        var (mPrime, ok) = TensorCode.Decode(vMinusUY, p);
        if (!ok || mPrime == null)
        {
            mPrime = new byte[p.VecKSizeBytes];
            Array.Copy(sigma, mPrime, p.VecKSizeBytes);
        }

        // Re-encrypt to verify
        byte[] thetaPrime = ComputeTheta(mPrime, pk, d, p);
        var (u2, v2) = PkeEncrypt(mPrime, thetaPrime, pk, p);

        // Constant-time comparison
        ulong[] u2Trunc = GF2.VectResize(u2, p.N);
        ulong[] uOrig = GF2.VectResize(u, p.N);
        int uMatch = GF2.VectEqual(u2Trunc, uOrig);

        ulong[] v2Trunc = GF2.VectResize(v2, p.N1N2);
        ulong[] vOrig = GF2.VectResize(v, p.N1N2);
        int vMatch = GF2.VectEqual(v2Trunc, vOrig);

        int match = uMatch & vMatch;

        // Constant-time selection of message or sigma
        byte[] mc = new byte[p.VecKSizeBytes];
        byte maskOK = (byte)(0 - (byte)match);       // 0xFF if match, 0x00 otherwise
        byte maskFail = (byte)(0 - (byte)(1 - match)); // 0x00 if match, 0xFF otherwise
        for (int i = 0; i < p.VecKSizeBytes; i++)
            mc[i] = (byte)((mPrime[i] & maskOK) | (sigma[i] & maskFail));

        return ComputeSS(mc, u, v, p);
    }

    // --- Private helpers ---

    private static (ulong[] u, ulong[] v) PkeEncrypt(byte[] m, byte[] theta, byte[] pk, HqcParams p)
    {
        // Parse public key
        byte[] pkSeed = pk[..HqcParams.SeedBytes];
        ulong[] s = GF2.VectFromBytes(pk.AsSpan(HqcParams.SeedBytes), p.VecNSize64);

        // Generate h from pk_seed
        var pkExpander = new SeedExpander(pkSeed);
        ulong[] h = VectSetRandom(pkExpander, p.N);

        // Generate r1, r2, e from theta
        var thetaExpander = new SeedExpander(theta);
        ulong[] r1 = VectSetRandomFixedWeight(thetaExpander, p.N, p.WR);
        ulong[] r2 = VectSetRandomFixedWeight(thetaExpander, p.N, p.WR);
        ulong[] e = VectSetRandomFixedWeight(thetaExpander, p.N, p.WE);

        // u = r1 + h * r2 mod (x^n - 1)
        ulong[] hr2 = GF2.VectMul(h, r2, p.N);
        ulong[] u = GF2.VectAdd(hr2, r1);
        u = GF2.VectResize(u, p.N);

        // v = encode(m) + s * r2 + e
        ulong[] encoded = TensorCode.Encode(m, p);

        ulong[] sr2 = GF2.VectMul(s, r2, p.N);
        var sr2Trunc = new ulong[p.VecN1N2Size64];
        Array.Copy(sr2, sr2Trunc, Math.Min(sr2.Length, p.VecN1N2Size64));
        if (p.N1N2 % 64 != 0 && p.VecN1N2Size64 > 0)
            sr2Trunc[p.VecN1N2Size64 - 1] &= (1UL << (p.N1N2 % 64)) - 1;

        var eResized = new ulong[p.VecN1N2Size64];
        Array.Copy(e, eResized, Math.Min(e.Length, p.VecN1N2Size64));
        if (p.N1N2 % 64 != 0 && p.VecN1N2Size64 > 0)
            eResized[p.VecN1N2Size64 - 1] &= (1UL << (p.N1N2 % 64)) - 1;

        ulong[] v = GF2.VectAdd(encoded, sr2Trunc);
        v = GF2.VectAdd(v, eResized);
        v = GF2.VectResize(v, p.N1N2);

        return (u, v);
    }

    /// <summary>d = SHAKE256(H_domain || m), 64 bytes.</summary>
    private static byte[] ComputeD(byte[] m)
    {
        byte[] input = new byte[1 + m.Length];
        input[0] = HqcParams.HFctDomain;
        Array.Copy(m, 0, input, 1, m.Length);
        return Shake256Impl.Hash(input, HqcParams.SharedSecretBytes);
    }

    /// <summary>theta = SHAKE256(G_domain || m || pk || d), SeedBytes output.</summary>
    private static byte[] ComputeTheta(byte[] m, byte[] pk, byte[] d, HqcParams p)
    {
        byte[] input = new byte[1 + m.Length + pk.Length + d.Length];
        input[0] = HqcParams.GFctDomain;
        int off = 1;
        Array.Copy(m, 0, input, off, m.Length); off += m.Length;
        Array.Copy(pk, 0, input, off, pk.Length); off += pk.Length;
        Array.Copy(d, 0, input, off, d.Length);
        return Shake256Impl.Hash(input, HqcParams.SeedBytes);
    }

    /// <summary>ss = SHAKE256(K_domain || m || u_bytes || v_bytes), 64 bytes.</summary>
    private static byte[] ComputeSS(byte[] m, ulong[] u, ulong[] v, HqcParams p)
    {
        byte[] uBytes = GF2.VectToBytes(u, p.VecNSizeBytes);
        byte[] vBytes = GF2.VectToBytes(v, p.VecN1N2SizeBytes);
        byte[] input = new byte[1 + m.Length + uBytes.Length + vBytes.Length];
        input[0] = HqcParams.KFctDomain;
        int off = 1;
        Array.Copy(m, 0, input, off, m.Length); off += m.Length;
        Array.Copy(uBytes, 0, input, off, uBytes.Length); off += uBytes.Length;
        Array.Copy(vBytes, 0, input, off, vBytes.Length);
        return Shake256Impl.Hash(input, HqcParams.SharedSecretBytes);
    }

    /// <summary>Generates a random vector of n bits using a seed expander.</summary>
    private static ulong[] VectSetRandom(SeedExpander se, int n)
    {
        int nWords = (n + 63) / 64;
        int nBytes = nWords * 8;
        byte[] buf = new byte[nBytes];
        se.Read(buf);
        ulong[] v = GF2.VectFromBytes(buf, nWords);
        int rem = n % 64;
        if (rem != 0)
            v[nWords - 1] &= (1UL << rem) - 1;
        return v;
    }

    /// <summary>
    /// Generates a random vector of n bits with exactly 'weight' bits set,
    /// using the seed expander with rejection sampling.
    /// </summary>
    private static ulong[] VectSetRandomFixedWeight(SeedExpander se, int n, int weight)
    {
        int nWords = (n + 63) / 64;
        var v = new ulong[nWords];
        var positions = new uint[weight];
        byte[] buf = new byte[4];

        for (int i = 0; i < weight; i++)
        {
            while (true)
            {
                se.Read(buf);
                uint pos = (uint)buf[0] | ((uint)buf[1] << 8) | ((uint)buf[2] << 16) | ((uint)buf[3] << 24);
                pos %= (uint)n;

                bool duplicate = false;
                for (int j = 0; j < i; j++)
                {
                    if (positions[j] == pos)
                    {
                        duplicate = true;
                        break;
                    }
                }
                if (!duplicate)
                {
                    positions[i] = pos;
                    break;
                }
            }
        }

        foreach (uint pos in positions)
            GF2.VectSetBit(v, (int)pos);

        return v;
    }

    /// <summary>SHAKE-256 based seed expander (streaming XOF).</summary>
    internal sealed class SeedExpander
    {
        // We implement incremental SHAKE-256 squeeze using Keccak state.
        // Absorb the seed once, then squeeze as needed.
        private readonly ulong[] _state = new ulong[25];
        private readonly byte[] _rateBuf = new byte[136]; // rate = 136 for SHAKE-256
        private int _offset = 136; // start with empty buffer to force first squeeze

        public SeedExpander(byte[] seed)
        {
            // Absorb seed using SHAKE-256 sponge
            AbsorbSeed(seed);
        }

        private void AbsorbSeed(byte[] seed)
        {
            const int rate = 136;
            const byte domainSuffix = 0x1F; // SHAKE-256

            Array.Clear(_state);
            int offset = 0;
            int remaining = seed.Length;

            // Absorb full blocks
            while (remaining >= rate)
            {
                AbsorbBlock(seed.AsSpan(offset, rate));
                KeccakF1600();
                offset += rate;
                remaining -= rate;
            }

            // Absorb last block with padding
            var lastBlock = new byte[rate];
            if (remaining > 0)
                Array.Copy(seed, offset, lastBlock, 0, remaining);
            lastBlock[remaining] = domainSuffix;
            lastBlock[rate - 1] |= 0x80;
            AbsorbBlock(lastBlock);
            KeccakF1600();

            // Squeeze first rate-bytes into buffer
            SqueezeBlock();
            _offset = 0;
        }

        private void AbsorbBlock(ReadOnlySpan<byte> data)
        {
            int laneCount = data.Length / 8;
            for (int i = 0; i < laneCount; i++)
                _state[i] ^= System.Buffers.Binary.BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(i * 8, 8));
            int tail = data.Length % 8;
            if (tail > 0)
            {
                ulong lane = 0;
                for (int b = 0; b < tail; b++)
                    lane |= (ulong)data[laneCount * 8 + b] << (8 * b);
                _state[laneCount] ^= lane;
            }
        }

        private void SqueezeBlock()
        {
            const int rate = 136;
            int laneCount = rate / 8;
            for (int i = 0; i < laneCount; i++)
                System.Buffers.Binary.BinaryPrimitives.WriteUInt64LittleEndian(_rateBuf.AsSpan(i * 8, 8), _state[i]);
        }

        public void Read(byte[] output) => Read(output.AsSpan());

        public void Read(Span<byte> output)
        {
            const int rate = 136;
            int needed = output.Length;
            int dstOff = 0;

            while (needed > 0)
            {
                int avail = rate - _offset;
                if (avail <= 0)
                {
                    KeccakF1600();
                    SqueezeBlock();
                    _offset = 0;
                    avail = rate;
                }
                int toCopy = Math.Min(avail, needed);
                _rateBuf.AsSpan(_offset, toCopy).CopyTo(output[dstOff..]);
                _offset += toCopy;
                dstOff += toCopy;
                needed -= toCopy;
            }
        }

        // Keccak-f[1600] permutation (same as Shake256Impl)
        private static readonly ulong[] RoundConstants =
        [
            0x0000000000000001UL, 0x0000000000008082UL,
            0x800000000000808AUL, 0x8000000080008000UL,
            0x000000000000808BUL, 0x0000000080000001UL,
            0x8000000080008081UL, 0x8000000000008009UL,
            0x000000000000008AUL, 0x0000000000000088UL,
            0x0000000080008009UL, 0x000000008000000AUL,
            0x000000008000808BUL, 0x800000000000008BUL,
            0x8000000000008089UL, 0x8000000000008003UL,
            0x8000000000008002UL, 0x8000000000000080UL,
            0x000000000000800AUL, 0x800000008000000AUL,
            0x8000000080008081UL, 0x8000000000008080UL,
            0x0000000080000001UL, 0x8000000080008008UL,
        ];

        private static readonly int[] RotationOffsets =
        [
             0,  1, 62, 28, 27,
            36, 44,  6, 55, 20,
             3, 10, 43, 25, 39,
            41, 45, 15, 21,  8,
            18,  2, 61, 56, 14,
        ];

        private static readonly int[] PiLane =
        [
             0, 10, 20,  5, 15,
            16,  1, 11, 21,  6,
             7, 17,  2, 12, 22,
            23,  8, 18,  3, 13,
            14, 24,  9, 19,  4,
        ];

        private void KeccakF1600()
        {
            Span<ulong> C = stackalloc ulong[5];
            Span<ulong> temp = stackalloc ulong[25];

            for (int round = 0; round < 24; round++)
            {
                for (int x = 0; x < 5; x++)
                    C[x] = _state[x] ^ _state[x + 5] ^ _state[x + 10] ^ _state[x + 15] ^ _state[x + 20];

                for (int x = 0; x < 5; x++)
                {
                    ulong dd = C[(x + 4) % 5] ^ RotateLeft(C[(x + 1) % 5], 1);
                    for (int yy = 0; yy < 25; yy += 5)
                        _state[yy + x] ^= dd;
                }

                for (int i = 0; i < 25; i++)
                    temp[PiLane[i]] = RotateLeft(_state[i], RotationOffsets[i]);

                for (int yy = 0; yy < 25; yy += 5)
                {
                    for (int x = 0; x < 5; x++)
                        _state[yy + x] = temp[yy + x] ^ (~temp[yy + (x + 1) % 5] & temp[yy + (x + 2) % 5]);
                }

                _state[0] ^= RoundConstants[round];
            }
        }

        private static ulong RotateLeft(ulong value, int offset)
            => (value << offset) | (value >> (64 - offset));
    }
}

File: swift/hqc/Sources/Hqc/HqcKem.swift
// HqcKem.swift -- HQC KEM: Key Generation, Encapsulation, Decapsulation.
//
// Public API following the Go reference implementation structure.

import Foundation

// MARK: - Seed Expander

/// SHAKE256-based seed expander.
final class SeedExpander {
    private let shake: SHAKE256

    init(_ seed: [UInt8]) {
        shake = SHAKE256()
        shake.absorb(seed)
    }

    func read(_ count: Int) -> [UInt8] {
        return shake.squeeze(count)
    }
}

// MARK: - Random vector generation

/// Generate a random vector of n bits using the seed expander.
func vectSetRandom(_ se: SeedExpander, _ n: Int) -> [UInt64] {
    let nWords = (n + 63) / 64
    let nBytes = nWords * 8
    let buf = se.read(nBytes)
    var v = vectFromBytes(buf, nWords)
    let rem = n % 64
    if rem != 0 {
        v[nWords - 1] &= (1 << rem) - 1
    }
    return v
}

/// Generate a random vector of n bits with exactly `weight` bits set.
/// Uses rejection sampling from sequential 4-byte chunks.
func vectSetRandomFixedWeight(_ se: SeedExpander, _ n: Int, _ weight: Int) -> [UInt64] {
    let nWords = (n + 63) / 64
    var v = [UInt64](repeating: 0, count: nWords)

    var positions = [UInt32](repeating: 0, count: weight)

    for i in 0..<weight {
        while true {
            let buf = se.read(4)
            let pos = (UInt32(buf[0]) | UInt32(buf[1]) << 8 |
                       UInt32(buf[2]) << 16 | UInt32(buf[3]) << 24) % UInt32(n)

            // Check for duplicates
            var duplicate = false
            for j in 0..<i {
                if positions[j] == pos {
                    duplicate = true
                    break
                }
            }
            if !duplicate {
                positions[i] = pos
                break
            }
        }
    }

    for pos in positions {
        vectSetBit(&v, Int(pos))
    }

    return v
}

// MARK: - Hash functions

/// Compute d = SHAKE256(H_domain || m), producing 64 bytes.
func computeD(_ m: [UInt8]) -> [UInt8] {
    let h = SHAKE256()
    h.absorb([hFctDomain])
    h.absorb(m)
    return h.squeeze(sharedSecretBytes)
}

/// Compute theta = SHAKE256(G_domain || m || pk || d).
func computeTheta(_ m: [UInt8], _ pk: [UInt8], _ d: [UInt8], _ p: HqcParams) -> [UInt8] {
    let h = SHAKE256()
    h.absorb([gFctDomain])
    h.absorb(m)
    h.absorb(pk)
    h.absorb(d)
    return h.squeeze(seedBytes)
}

/// Compute ss = SHAKE256(K_domain || m || u_bytes || v_bytes).
func computeSS(_ m: [UInt8], _ u: [UInt64], _ v: [UInt64], _ p: HqcParams) -> [UInt8] {
    let h = SHAKE256()
    h.absorb([kFctDomain])
    h.absorb(m)
    h.absorb(vectToBytes(u, p.vecNSizeBytes))
    h.absorb(vectToBytes(v, p.vecN1N2SizeBytes))
    return h.squeeze(sharedSecretBytes)
}

// MARK: - PKE Encrypt

/// Internal PKE encryption.
func pkeEncrypt(_ m: [UInt8], _ theta: [UInt8], _ pk: [UInt8],
                _ p: HqcParams) -> ([UInt64], [UInt64]) {
    // Parse public key
    let pkSeed = Array(pk[0..<seedBytes])
    let s = vectFromBytes(Array(pk[seedBytes...]), p.vecNSize64)

    // Generate h from pk_seed
    let pkExpander = SeedExpander(pkSeed)
    let h = vectSetRandom(pkExpander, p.n)

    // Generate r1, r2 with weight WR and e with weight WE from theta
    let thetaExpander = SeedExpander(theta)
    let r1 = vectSetRandomFixedWeight(thetaExpander, p.n, p.wr)
    let r2 = vectSetRandomFixedWeight(thetaExpander, p.n, p.wr)
    let e = vectSetRandomFixedWeight(thetaExpander, p.n, p.we)

    // u = r1 + h * r2 mod (x^n - 1)
    let hr2 = vectMul(h, r2, p.n)
    var u = vectAdd(hr2, r1)
    u = vectResize(u, p.n)

    // v = encode(m) + s * r2 + e
    let encoded = tensorEncode(m, p)

    // s * r2 in the ring, then truncate to n1*n2 bits
    let sr2 = vectMul(s, r2, p.n)
    var sr2Trunc = [UInt64](repeating: 0, count: p.vecN1N2Size64)
    for i in 0..<min(sr2.count, p.vecN1N2Size64) { sr2Trunc[i] = sr2[i] }
    if p.n1n2 % 64 != 0 && p.vecN1N2Size64 > 0 {
        sr2Trunc[p.vecN1N2Size64 - 1] &= (1 << (p.n1n2 % 64)) - 1
    }

    // Resize e to n1*n2
    var eResized = [UInt64](repeating: 0, count: p.vecN1N2Size64)
    for i in 0..<min(e.count, p.vecN1N2Size64) { eResized[i] = e[i] }
    if p.n1n2 % 64 != 0 && p.vecN1N2Size64 > 0 {
        eResized[p.vecN1N2Size64 - 1] &= (1 << (p.n1n2 % 64)) - 1
    }

    var v = vectAdd(encoded, sr2Trunc)
    v = vectAdd(v, eResized)
    v = vectResize(v, p.n1n2)

    return (u, v)
}

// MARK: - Public API

/// Generate an HQC key pair.
/// Returns (publicKey, secretKey).
public func hqcKeyGen(_ p: HqcParams) -> ([UInt8], [UInt8]) {
    initGF256Tables()

    // Generate random seeds
    var skSeed = [UInt8](repeating: 0, count: seedBytes)
    var pkSeed = [UInt8](repeating: 0, count: seedBytes)
    _ = SecRandomCopyBytes(kSecRandomDefault, seedBytes, &skSeed)
    _ = SecRandomCopyBytes(kSecRandomDefault, seedBytes, &pkSeed)

    // Generate secret vectors x, y from sk_seed
    let skExpander = SeedExpander(skSeed)
    let x = vectSetRandomFixedWeight(skExpander, p.n, p.w)
    let y = vectSetRandomFixedWeight(skExpander, p.n, p.w)

    // Generate random vector h from pk_seed
    let pkExpander = SeedExpander(pkSeed)
    let h = vectSetRandom(pkExpander, p.n)

    // Compute s = x + h * y mod (x^n - 1)
    let hy = vectMul(h, y, p.n)
    var s = vectAdd(hy, x)
    s = vectResize(s, p.n)

    // Serialize public key: [pk_seed (40 bytes)] [s (VecNSizeBytes bytes)]
    var pk = [UInt8](repeating: 0, count: p.pkSize)
    for i in 0..<seedBytes { pk[i] = pkSeed[i] }
    let sBytes = vectToBytes(s, p.vecNSizeBytes)
    for i in 0..<sBytes.count { pk[seedBytes + i] = sBytes[i] }

    // Serialize secret key: [sk_seed (40 bytes)] [pk (PKSize bytes)]
    var sk = [UInt8](repeating: 0, count: p.skSize)
    for i in 0..<seedBytes { sk[i] = skSeed[i] }
    for i in 0..<pk.count { sk[seedBytes + i] = pk[i] }

    return (pk, sk)
}

/// Encapsulate a shared secret using the public key.
/// Returns (ciphertext, sharedSecret).
public func hqcEncaps(_ pk: [UInt8], _ p: HqcParams) -> ([UInt8], [UInt8]) {
    initGF256Tables()

    // Generate random message m
    var m = [UInt8](repeating: 0, count: p.vecKSizeBytes)
    _ = SecRandomCopyBytes(kSecRandomDefault, p.vecKSizeBytes, &m)

    // Compute d = H(m)
    let d = computeD(m)

    // Compute theta = SHAKE256(G_domain || m || pk || d)
    let theta = computeTheta(m, pk, d, p)

    // PKE Encrypt
    let (u, v) = pkeEncrypt(m, theta, pk, p)

    // Compute shared secret
    let ss = computeSS(m, u, v, p)

    // Serialize ciphertext: [u (VecNSizeBytes)] [v (VecN1N2SizeBytes)] [d (64 bytes)]
    var ct = [UInt8](repeating: 0, count: p.ctSize)
    let uBytes = vectToBytes(u, p.vecNSizeBytes)
    let vBytes = vectToBytes(v, p.vecN1N2SizeBytes)
    for i in 0..<uBytes.count { ct[i] = uBytes[i] }
    for i in 0..<vBytes.count { ct[p.vecNSizeBytes + i] = vBytes[i] }
    for i in 0..<d.count { ct[p.vecNSizeBytes + p.vecN1N2SizeBytes + i] = d[i] }

    return (ct, ss)
}

/// Decapsulate a shared secret from a ciphertext using the secret key.
public func hqcDecaps(_ sk: [UInt8], _ ct: [UInt8], _ p: HqcParams) -> [UInt8] {
    initGF256Tables()

    // Parse secret key: [sk_seed (40 bytes)] [pk (PKSize bytes)]
    let skSeed = Array(sk[0..<seedBytes])
    let pk = Array(sk[seedBytes...])

    // Parse ciphertext: [u (VecNSizeBytes)] [v (VecN1N2SizeBytes)] [d (64 bytes)]
    let u = vectFromBytes(Array(ct[0..<p.vecNSizeBytes]), p.vecNSize64)
    let v = vectFromBytes(Array(ct[p.vecNSizeBytes..<(p.vecNSizeBytes + p.vecN1N2SizeBytes)]),
                          p.vecN1N2Size64)
    let d = Array(ct[(p.vecNSizeBytes + p.vecN1N2SizeBytes)...])

    // Regenerate secret vectors x, y and sigma from sk_seed
    let skExpander = SeedExpander(skSeed)
    let _ = vectSetRandomFixedWeight(skExpander, p.n, p.w)  // x (not needed for decrypt)
    let y = vectSetRandomFixedWeight(skExpander, p.n, p.w)
    // Generate sigma (rejection secret)
    let sigma = skExpander.read(p.vecKSizeBytes)

    // Compute v - u * y (which is v XOR u*y since we are in GF(2))
    let uy = vectMul(u, y, p.n)

    // Truncate uy to n1*n2 bits
    var uyTrunc = [UInt64](repeating: 0, count: p.vecN1N2Size64)
    for i in 0..<min(uy.count, p.vecN1N2Size64) { uyTrunc[i] = uy[i] }
    if p.n1n2 % 64 != 0 && p.vecN1N2Size64 > 0 {
        uyTrunc[p.vecN1N2Size64 - 1] &= (1 << (p.n1n2 % 64)) - 1
    }

    let vMinusUY = vectAdd(v, uyTrunc)

    // Decode using tensor product code
    var mPrime: [UInt8]
    let (decoded, ok) = tensorDecode(vMinusUY, p)
    if ok, let decodedMsg = decoded {
        mPrime = decodedMsg
    } else {
        // Decoding failed - use sigma
        mPrime = [UInt8](repeating: 0, count: p.vecKSizeBytes)
        for i in 0..<p.vecKSizeBytes { mPrime[i] = sigma[i] }
    }

    // Re-encrypt to verify
    let thetaPrime = computeTheta(mPrime, pk, d, p)
    let (u2, v2) = pkeEncrypt(mPrime, thetaPrime, pk, p)

    // Constant-time comparison
    let u2Trunc = vectResize(u2, p.n)
    let uOrig = vectResize(u, p.n)
    let uMatch = vectEqual(u2Trunc, uOrig)

    let v2Trunc = vectResize(v2, p.n1n2)
    let vOrig = vectResize(v, p.n1n2)
    let vMatch = vectEqual(v2Trunc, vOrig)

    let match = uMatch & vMatch

    // Constant-time selection of message or sigma
    var mc = [UInt8](repeating: 0, count: p.vecKSizeBytes)
    let maskOK = UInt8(0) &- UInt8(match)         // 0xFF if match, 0x00 otherwise
    let maskFail = UInt8(0) &- UInt8(1 - match)   // 0x00 if match, 0xFF otherwise
    for i in 0..<p.vecKSizeBytes {
        mc[i] = (mPrime[i] & maskOK) | (sigma[i] & maskFail)
    }

    // Compute shared secret
    let ss = computeSS(mc, u, v, p)
    return ss
}
